### DISCUSSION TITLE
Extend dataset manually with data of missing nuclide

### DISCUSSION THREAD

**Discussion by Cs137** [radioactivedecay/radioactivedecay#70] [Q&A] [answered]
Many thanks for the compilation and provision of this amazing package!

I wonder whether it is possible to add data for nuclides that are not included in the default dataset. Since you state that the default data are based on _ICRP Publication 107_, I assume that there is a possibility to make use of a custom dataset. Sadly I am not able to find information on how to use a custom dataset in the documentation. 

According to what I get from the documentation, I would tackle it as follows:

1. export the arguments required to initiate the [`DecayData`](https://radioactivedecay.github.io/decaydata.html#id1) instance, 
2. extend it with the missing information, and 
3. initialise a new instance using the modified arguments.

Would someone be so kind to give me a hint how to do that in practice?

**Comment by alexmalins** [comment, marked as answer]
Hi @Cs137 - thanks for raising this.

Yes, the documentation around creating and using datasets is an area that needs improving. Working on this in on the roadmap for release v0.5.

You are right - the code is designed so using custom decay datasets is possible. It is by instantiating a new `DecayData` instance as you correctly identified, and then supplying this instance to the `Inventory` or `Nuclide` constructors via the `decay_data=...` parameter.

Examples for creating the data structures/files needed to create `DecayData` instances are currently housed in the [datasets](https://github.com/radioactivedecay/datasets) repo in Jupyter notebooks. It is possible to follow these notebooks and extend to add data for missing nuclides you need. But I wouldn't necessarily recommend this, unless you are highly motivated, as I plan to change the underlying data structures in v0.5 and simplify everything.

How urgently do you need the missing nuclides? Are you are ok to wait a few weeks for v0.5 when everything will be better? Or if you need quicker, I could support you in creating custom dataset files for your problem now?

p.s. In the long-term we hope to incorporate datasets other than ICRP 107, e.g. ENDF/B-VIII.0, which contain more nuclides. This will come after release v0.5, however.

**Reply by Cs137** [reply]
Dear @alexmalins 

thank you very much for the response and suggesting a solution. I hope you did not get me wrong: My question was not meant to be a criticism of the documentation. I am very happy to be able to benefit of the work and time you've invested to create this package!

I do not need to solve the problem urgently, since the nuclides I am currently interested in are present in the _ICRP 107_ dataset. For a future project it would be of intereset. In particular, _Xe-136_ is missing. I looked up the [dataset repo](https://github.com/radioactivedecay/datasets) and recognised that it is part of the [pyne_full.csv](https://github.com/radioactivedecay/datasets/blob/main/pyne/pyne_full.csv) and [pyne_truncated.csv](https://github.com/radioactivedecay/datasets/blob/main/pyne/pyne_truncated.csv) files, but is missing in the [icrp.csv](https://github.com/radioactivedecay/datasets/blob/main/icrp107_ame2020_nubase2020/icrp.csv) file. So the information are provided by AME2020 and Nubase2020 and the usage of an alternative dataset would solve my issue, as it is included in _ENDF/B-VIII.0_.

Also many thanks for your offer to help me out, but I'll just be patient. If I can be of any help for your work besides mentioning it in my scientific output, I'll be glad to do so - best regards!

**Reply by Cs137** [reply]
Another unrelated issue: 
I recognised that the `yscale='log'` argument does result in a plot with logarithmic scaling when including it in the `Inventory.plot()` method, but no dataseries are plotted. I do not need it and it doesn't cause any troubles for me, but I assume that's not what it is supposed to do.

Since I am not so experienced with the github workflows and am usually more a kind of a silent reader, I mention it here - but can also create an issue if you want to keep track of it.

Cheers!

**Reply by alexmalins** [reply]
Thanks @Cs137

I guess ICRP-107 did not include Xe-136 as it has an extremely long half-life, so it is of little practical significance for radiation protection, which is what the ICRP-107 dataset was originally intended for.

Yes you could use the PyNE dataset files from [this folder](https://github.com/radioactivedecay/datasets/tree/main/pyne) to model Xe-136. To do this use the `rd.decaydata.load_dataset()` function described [here](https://radioactivedecay.github.io/decaydata.html). Simple example:

```
>>> pynedata= rd.decaydata.load_dataset(dataset_name='pyne', dir_path='path/to/where/you/downloaded/files', load_sympy=True)
>>> inv = rd.InventoryHP({'Xe-136': 1.0}, decay_data=pynedata)
>>> inv.decay(10, 'My')
```

Note PyNE does not include the stable progeny outcome of Xe-136 double beta decay (Ba-136). This and other issues are reasons why PyNE data isn't supplied automatically within radioactivedecay.

~~**Edit: Just realized the PyNE dataset files [here](https://github.com/radioactivedecay/datasets/tree/main/pyne) need to be updated to be compatible with radioactivedecay v0.4.5+ due to the change in SymPy pickle format. I'll try to do this later today or tomorrow. As a workaround if you want to try now, rename the end of the `.pickle` files to `_1.8.pickle`, and make sure SymPy v1.8 or less is in your environment.**~~

**Edit 2022/03/22: datasets repo now has PyNE dataset files compatible with  radioactivedecay v0.4.5+**

**Reply by alexmalins** [reply]
The issue with `yscale='log'` and nothing showing up is because the maximum amount (activity/moles/whatever you were plotting) of any of the nuclides in your case would have been lower than 0.1. The default `ymin` for a logscale y-axis was 0.1 before.

The fact that nothing was plotted was counter-intuitive, so I just changed the code so the default `ymin` for a log y-axis is 0.95x the minimum quantity of any nuclide over the decay period. This means data will now always show on the graph. In some circumstances `ymin` will be very small (e.g. 10^-300) with the new logic, but users can always override this by supplying an appropriate `ymin=...` to the `plot()` method themselves.

Please see the latest release (v0.4.11) and if it solves the issue you had.


Thanks for your offer of help. The best way you can support this project is to keep up what you are doing! Please don't be silent or assume we already know about any issues you find - please keep contributing on here 😊

**Comment by aairfccha** [comment]
Exhuming this thread: I fed the output of a burn-up calculation to radioactivedecay and got just 10 "unknown" nuclides: Ca-48, Ge-76, Se-82, Zr-96, Mo-100, Tc-100, Xe-136, Nd-150, Bk-248 and Es-252. Most of those have half-lives of 10 to the power of something two-digit years, Tc-100 is gone pretty fast (half-life of 15 seconds) and the last two are pretty oddball transuranics. So the missing nuclides are probably not much of a practical problem, just a bit of an annoyance.

**Reply by alexmalins** [reply]
Hi @aairfccha thanks for your comment and apologies for the slow reply. Yes, those nuclides are not in the ICRP-107 dataset. One day I hope to find the time to make another dataset for this code with better coverage, e.g. [SandiaDecay](https://github.com/sandialabs/SandiaDecay/tree/master)

Or if you are interested in trying to make a dataset yourself, take a look in the [datasets repo](https://github.com/radioactivedecay/datasets) for example notebooks of how that can be done (for ICRP-107 and PyNE decay data)

**Reply by aairfccha** [reply]
Through a lot of tinkering with endf-parserpy, I think I managed to wring most data required to build a nuclide dataset out of ENDF-files. Looping over all files of the ENDF-B/VII.1 set of decay data, this yielded a table from neutrons to Roentgenium-272 (3822 nuclides) with the following entries:

- short name of nuclide (like Ag-95o)
- Z
- A
- element abbreviation (currently not present but could be trivially added)
- excitation level
- atomic weight
- nuclear stability token
- Half-life
- thermally effective energies for light particles, heavy particles, EM-radiation and the sum of those
- number of decay paths provided
- mode of decay for each path in ENDF-notation
- branching fraction for each path
- excitation level of decay product for each path
- total energy for each path (probably useless for most practical applications as it includes energy of neutrinos)

The long names for each element are missing, but for this a manually generated table is viable and this should be the only external data needed. The decay products have to be calculated but the necessary information to do that should be there.

Next step is to make the procedure in the jupyter notebook work for this sort of file. No idea whether my python skills are up to that (and then there is finding the time...).

**Reply by aairfccha** [reply]
I braved parts of the byzantine process and with a lot of wrangling managed to generate a dataframe which at least appears to replicate the format of icrp.csv. Now there are two new roadblocks.

- Some nuclides in some ENDF data collections seem to have decay paths which lead to a nuclide not contained in the data collection. Am I assuming correctly that this is kind of bad for decay matrix generation? How best to deal with that, remove those nuclides before matrix generation?
- The notebook for processing the ICRP dataset mentions that two nuclides with identical half-lives can be a problem but the ICRP data doesn't have those. Unfortunately the ENDF data collections **do** have nuclides with colliding identical half-lives. How to deal with that, manually fudge one half-life of each pair by a tenth of a second? Doing this fully manually would be a solution I am not really keen on as it would require updating the script for a new or updated data source.
